### ISSUE TITLE
Distinguish types with the same name from different modules

### ISSUE THREAD

**Issue by ilyapuchka** [krzysztofzablocki/Sourcery#214]
Currently multiple targets support has a limitation that it is not able to distinguish types with the same name coming from different modules. If types with the same name from different modules will be scanned together they will be merged into type scanned first, as they are mapped by their name, which does not contain module name. Also there is no way to access such types in the templates.

**Comment by Antondomashnev** [comment]
As far as I can see from the code, we can close this issue, could you confirm it please?

**Comment by ilyapuchka** [comment]
Nope, it was never done and to be honest, I'm not sure how to approach this, especially in terms of templates api.

**Comment by Antondomashnev** [comment]
> If types with the same name from different modules will be scanned together they will be merged into type scanned first, as they are mapped by their name, which does not contain module name. 

Wasn't it resolved with `module` and `globalName` properties on `Type`? I can see in `Composer` that it has the `modules` dictionary to collect types by module name.

**Comment by ilyapuchka** [comment]
it collects, but I don't think it properly distinguish types mentioned in scanned code. Because if you have two types named the same from different modules what type we should assign to the declaration depends on the current scope of the file, what is being imported. Unless you use full type name. And other issues like how to mention these types in the templates. If you mention it by type name without module which one we should return? Currently we just override one type with another in types collections.

**Comment by Antondomashnev** [comment]
> If you mention it by type name without module which one we should return? Currently we just override one type with another in types collections.

I think this should be done on the client's shoulders. So in the template there should be somehow information which module to use. But I got the point that this is not done yet.

**Comment by ilyapuchka** [comment]
@rubensamsonyanauto1 I believe what you described is a different problem, can you please create a new issue for it?

**Comment by MaximKotliar** [comment]
Any updates on this issue?

**Comment by krzysztofzablocki** [comment]
Sourcery have been doing lookups by module for a while now but there will be big changes in next releases anyway so if it doesn't do something atm then it won't until that AST rewrite happens

**Comment by schmiter** [comment]
Are there any updates on this issue?
